casa/IO/MultiFile.cc

```diff
@@ -36,191 +37,752 @@
+    }
+  }
+
+  void MultiFile::storeCRC (const void* buffer, Int64 blknr)
+  {
+    if (blknr >= Int64(itsCRC.size())) {
+      itsCRC.resize (blknr+1);
+    }
+    itsCRC[blknr] = calcCRC (buffer, itsBlockSize);
+  }
+
+  void MultiFile::checkCRC (const void* buffer, Int64 blknr) const
+  {
+    AlwaysAssert (blknr < Int64(itsCRC.size()), AipsError);
+    uInt crc = calcCRC (buffer, itsBlockSize);
+    if (crc != itsCRC[blknr]) {
+      throw AipsError ("Mismatch in CRC of MultiFile block " +
+                       String::toString(blknr));
+    }
+  }
+
+  uInt MultiFile::calcCRC (const void* buffer, Int64 size) const
+  {
+    // The algorithm is taken from crctester.c by Sven Reifegerste
+    // (www.zorc/reflex)
+    // See also https://www.lammertbies.nl/comm/info/crc-calculation
+    const uChar* buf = static_cast<const uChar*>(buffer);
+    const uInt crcinit = 0x46af6449;
+    //# The above value is calculated from:
+    //#         const uInt highbit = (uInt)1<<(32-1);
+    //#		crcinit = 0xffffffff;
+    //#		for (i=0; i<32; i++) {
+    //#			bit = crcinit & 1;
+    //#			if (bit) crcinit^= polynom;
+    //#			crcinit >>= 1;
+    //#			if (bit) crcinit|= highbit;
+    //#		}	
+    // Normal lookup table algorithm with augmented zero bytes.
+    uInt crc = crcinit;
+    for (Int64 i=0; i<size; ++i) {
+      crc = ((crc << 8) | buf[i]) ^ itsCRCTab[ (crc >> (32-8))  & 0xff];
+    }
+    // Augment zero bytes.
+    for (size_t i=0; i<32/8; ++i) {
+      crc = (crc << 8) ^ itsCRCTab[ (crc >> (32-8))  & 0xff];
+    }
+    crc^= 0xffffffff;
+    return crc;
+  }
+
+  void MultiFile::fillCRCTable()
+  {
+    // CRC-32 parameters
+    // 'polynom' is the standard CRC-32 polynom without leading '1' bit
+    const uInt polynom = 0x4c11db7;
+    const uInt highbit = (uInt)1<<(32-1);
+    // make CRC lookup table used by table algorithms
+    for (int i=0; i<256; i++) {
+      uInt crc = i;
+      crc<<= 32-8;
+      for (int j=0; j<8; j++) {
+        uInt bit = crc & highbit;
+        crc<<= 1;
+        if (bit) crc^= polynom;
+      }			
+      itsCRCTab[i]= crc;
+    }
+  }
+
+  void MultiFile::show (std::ostream& os) const
+  {
+    os << fileName() << ": blocksize=" << blockSize()
+       << "  nfile="  << nfile() << "  nblock=" << nblock()
+       << "  nCRC=" << itsCRC.size() << endl
+       << "  ncont=" << itsNrContUsed[0] << ',' << itsNrContUsed[1]
+       << "  cont=" << itsHdrContInx
+       << ' ' << itsHdrCont[itsHdrContInx].blockNrs
+       << "  free=" << freeBlocks() << endl;
+  }
+
+  vector<Int64> MultiFile::packIndex (const vector<Int64>& blockNrs)
+  {
+    // See if subsequent block numbers are used, so the index can be
+    // compressed.
+    // Compression is done by counting the nr of subsequent blocknrs
+    // and writing the negative count if > 0.
+    // Note that the count does not include the first block number.
+    vector<Int64> buf;
+    buf.reserve (blockNrs.size());
```

Comment:
let's move this after the if.

Reply:
and return vector<Int64() in the if?

Reply:
yes, `std::vector<Int64>()`

Reply:
Changed accordingly.  But what is the advantage? The reserve will do nothing because size==0.